### PR TITLE
Add offline meta-testing and per-task logging to MAML

### PR DESCRIPTION
**Pull request by naeioi**

### CODE REVIEW

**Comment by naeioi** [comment]
@Mergifyio rebase

File: examples/torch/maml_trpo_ml10_meta_test.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python3
+"""This is an example to do meta-testing on ML10 environment."""
+from metaworld.benchmarks import ML10
+
+from garage.experiment.offline_meta_test_helper \
```
Comment (ryanjulian) [inline examples/torch/maml_trpo_ml10_meta_test.py]: Please put this into the `garage.experiment` namespace

File: src/garage/_functions.py [hunk 2]
```diff
@@ -1,10 +1,71 @@
...
+            log NaN as a placeholder.
+
+    Returns:
+        numpy.ndarray: Undiscounted returns.
+
+    """
+    traj_by_name = defaultdict(list)
+    for trajectory in batch.split():
+        try:
+            task_name = trajectory.env_infos['task_name'][0]
```
Comment (ryanjulian) [inline src/garage/_functions.py]: can you leave an issue about unifying this interface?

File: src/garage/_functions.py [hunk 3]
```diff
@@ -1,10 +1,71 @@
...
+            if task_name in traj_by_name:
+                trajectories = traj_by_name[task_name]
+                log_performance(
+                    itr,
+                    garage.TrajectoryBatch.concatenate(*trajectories),
+                    discount,
+                    prefix=task_name)
+            else:
+                with tabular.prefix(task_name + '/'):
+                    tabular.record('Iteration', np.nan)
```
Comment (ryanjulian) [inline src/garage/_functions.py]: can you leave an issue on dowel about handling this case better?

File: src/garage/envs/task_id_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    def step(self, action):
+        """gym.Env step for the active task env.
+
+        Args:
+            action (np.ndarray): Action performed by the agent in the
+                environment.
+
+        Returns:
+            tuple:
```
Comment (ryanjulian) [inline src/garage/envs/task_id_wrapper.py]: returning a tuple is equivalent to multiple return values. please use that style instead (it looks just like the Args section)

File: src/garage/envs/task_id_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+"""Wrapper for Metaworld environment to provide properly-named tasks."""
+
+import gym
+
+
+class TaskIdWrapper(gym.Wrapper):
```
Comment (ryanjulian) [inline src/garage/envs/task_id_wrapper.py]: if this only works for metaworld, please put that in the name `MetaworldTaskIdWrapper`

File: src/garage/experiment/offline_meta_test_helper.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
+"""Offline meta-testing helper."""
```
Comment (ryanjulian) [inline src/garage/experiment/offline_meta_test_helper.py]: this file looks to me like it should live in scripts/

File: src/garage/torch/algos/_utils.py [hunk 7]
```diff
@@ -110,7 +110,7 @@ def compute_advantages(discount, gae_lambda, max_path_length, baselines,
     deltas = (rewards + discount * F.pad(baselines, (0, 1))[:, 1:] - baselines)
     deltas = F.pad(deltas, (0, max_path_length - 1)).unsqueeze(0).unsqueeze(0)
 
-    advantages = F.conv2d(deltas, adv_filter, stride=1).squeeze()
+    advantages = F.conv2d(deltas, adv_filter, stride=1).view(rewards.shape)
```
Comment (ryanjulian) [inline src/garage/torch/algos/_utils.py]: i also saw an update to this in #1188. Is this a bug?

Please confer with @yonghyuc and perhaps open a separate issue/PR so that it can be backported.

File: tests/garage/experiment/test_task_sampler.py [hunk 8]
```diff
@@ -6,6 +6,20 @@
 
 from garage.envs.half_cheetah_vel_env import HalfCheetahVelEnv
 from garage.experiment import task_sampler
+from garage.experiment.task_sampler import FullSetTaskSampler
+
+
+@pytest.mark.large
+def test_round_robin_set_task_sampler():
```
Comment (ryanjulian) [inline tests/garage/experiment/test_task_sampler.py]: do we really need to load ML10 to test this??

File: tests/garage/torch/misc/test_algo_utils.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
+"""Tests for utilities in torch/algos/_utils.py."""
+import torch
+
+from garage.torch.algos import compute_advantages
+
+
+class TestAlgoUtils:
```
Comment (ryanjulian) [inline tests/garage/torch/misc/test_algo_utils.py]: just an FYI that pytest doesn't require you define a class for this (you can just use functions called `test_...`)

File: examples/torch/maml_trpo_ml10_meta_test.py (hunk 1 above)
Comment (ryanjulian) [reply examples/torch/maml_trpo_ml10_meta_test.py]: please split lines with (), per PEP8

File: src/garage/torch/algos/_utils.py (hunk 7 above)
Comment (naeioi) [reply src/garage/torch/algos/_utils.py]: Yes, this is a bug that breaks when `rewards` has a shape of something like `(1, N)`. The expected shape of `advantages` is the same of that of `rewards`, but the `Tensor.seqeeze()` will squeeze out the the first dimension in this case.

Comment (naeioi) [reply src/garage/torch/algos/_utils.py]: @yonghyuc Because this is a bug that impacts old release of garage, you need to make a backport PR for this. If you don't know how to do it, here is an example https://github.com/rlworkgroup/garage/pull/1079.

**Comment by naeioi** [comment]
@ryanjulian OK. I will break this into several smaller PRs.

File: src/garage/torch/algos/maml.py [hunk 10]
```diff
@@ -419,34 +467,30 @@ class MAMLTrajectoryBatch(
...
     +-----------------------+-------------------------------------------------+
-    | :math:`N \bullet [T]` | A dimension computed by flattening a            |
-    |                       | variable-length time dimension :math:`[T]` into |
-    |                       | a single batch dimension with length            |
-    |                       | :math:`sum_{i \in N} [T]_i`                     |
-    +-----------------------+-------------------------------------------------+
 
     Attributes:
+        paths (list[dict[str, np.ndarray or dict[str, np.ndarray]]]):
+            Nonflatten original paths from sampler.
```
Comment (krzentner) [inline src/garage/torch/algos/maml.py]: Wait, why is this necessary? It kinda defeats the point of the datastructure, and doesn't seem to be read anywhere.

File: src/garage/_functions.py [hunk 11]
```diff
@@ -1,10 +1,71 @@
...
+from collections import defaultdict
+
 from dowel import tabular
 import numpy as np
 
+import garage
 from garage.misc.tensor_utils import discount_cumsum
 
 
+def log_multitask_performance(itr, batch, discount, task_names=None):
```
[the remark this answers is not in the document]
Comment (naeioi) [reply src/garage/_functions.py]: Sure! Will do.

File: src/garage/torch/algos/_utils.py (hunk 7 above)
Comment (naeioi) [reply src/garage/torch/algos/_utils.py]: Backport PR opened at #1197

File: src/garage/_functions.py (hunk 3 above)
Comment (naeioi) [reply src/garage/_functions.py]: Opened an issue at https://github.com/rlworkgroup/dowel/issues/45

File: src/garage/_functions.py (hunk 2 above)
Comment (naeioi) [reply src/garage/_functions.py]: Opened at https://github.com/rlworkgroup/garage/issues/1198

File: src/garage/experiment/offline_meta_test_helper.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+                        sys.exit()
+                    else:
+                        # In parent process
+                        children.append(pid)
+
+                for child in children:
+                    os.waitpid(child, 0)
+
+            if to_merge:
+                self._merge_csv(meta_train_dir, itrs)
```
Comment (krzentner) [inline src/garage/experiment/offline_meta_test_helper.py]: This file should probably be in `scripts/`, and should have a `if __name__ == '__main__'` block.

File: src/garage/envs/task_id_wrapper.py (hunk 5 above)
Comment (krzentner) [inline src/garage/envs/task_id_wrapper.py]: Wait, so is this class still necessary?
